### PR TITLE
Move stream management to transport

### LINKED ISSUE #12010: Move stream management to transport module
**Description**

In #11707, we added server side stream management api and implementation in the broker. 

Alternate is to add streams as a feature provided by the transport. Advantages of doing this is we can implement both server(broker) and client (gateway) side in one location. This will help us to unit test the feature without fully integrating it with the broker. On the broker side, we will only implement job specific implementation of the stream. 

The exact interface between transport and broker will be defined while working on this issue. We have already made the implementation related to stream-management job-agnostic to minimize the dependency between engine and platform.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

- Moved remote stream mangement api and its implementation from Broker to transport module. The goal is that the implementations for gateway side stream will be also added to transport. Having both client and server in the same module will help us to test the feature without fully integrating with the broker and gateway.
- Moved all sbe message definitions related to stream from protocol to transport. Since all consumers of these messages will be implemented in the transport module, there is no need to define it in the protocol which is a semi-public module.
- Moved message topic names to transport. To keep it simple, we assume there is only one type of stream (i.e all streams are job streams) will be created. So we don't support different topic names as of now.

While doing the refactoring, we noticed following potential refactoring. They are NOT included in this PR. They will be done in a separate PR. 
- Flatten `GatewayStreamer<M,P>` defined in stream-platform engine api to make it less generic. Eg:- JobGatewayStream without type parameters.
- Move this interface and related ones to the engine instead of stream-platform.
- Inject `GatewayStreamer` to engine via TypedRecordFactory, instead via stream-platform to remove the knowledge of "job activation" from the platform.
- Make `RemoteStreamMetrics` an interface so that metrics can be job specific and injected from the broker.

## Related issues

closes #12010

### CODE REVIEW

**Comment by deepthidevaki** [comment]
@npepinpe As we discussed, `LongPollingActivatesJobsTest` (IT)  failes when we did not impement notification in `JobGatewayStreamer`. So I didn't add a unit test for it.

File: transport/src/main/resources/stream-protocol.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  ~ with this work for additional information regarding copyright ownership.
+  ~ Licensed under the Zeebe Community License 1.1. You may not use this file
+  ~ except in compliance with the Zeebe Community License 1.1.
+  -->
+<sbe:messageSchema xmlns:sbe="http://fixprotocol.io/2016/sbe"
+  xmlns:xi="http://www.w3.org/2001/XInclude" package="io.camunda.zeebe.transport.stream.impl.messages"
+  id="0" version="1" semanticVersion="${project.version}"
+  description="Zeebe Protocol" byteOrder="littleEndian">
+
+  <xi:include href="../../../protocol/src/main/resources/common-types.xml"/>
```
Comment (npepinpe) [inline transport/src/main/resources/stream-protocol.xml:14]: :thought_balloon: I wish we didn't have to keep generating them over and over and over :smile:

File: transport/src/main/java/io/camunda/zeebe/transport/stream/impl/messages/StreamTopics.java [hunk 2]
```diff
@@ -5,18 +5,17 @@
...
-package io.camunda.zeebe.protocol.impl.stream;
+package io.camunda.zeebe.transport.stream.impl.messages;
 
-public enum JobStreamTopics {
-  ADD("job-stream-add"),
-  PUSH("job-stream-push"),
-  REMOVE("job-stream-remove"),
-  REMOVE_ALL("job-stream-remove-all"),
-  JOB_AVAILABLE("jobsAvailable");
+public enum StreamTopics {
```
Comment (npepinpe) [inline transport/src/main/java/io/camunda/zeebe/transport/stream/impl/messages/StreamTopics.java:10]: :thought_balloon: I wish we had some way of encoding those in the protocol as well :shrug:

File: transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamerImpl.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            StreamTopics.PUSH.topic(),
+            request,
+            this::serialize,
+            Function.identity(),
+            receiver,
+            REQUEST_TIMEOUT)
+        .thenApply(ok -> null);
+  }
+
+  private byte[] serialize(final BufferWriter payload) {
```
Comment (npepinpe) [inline transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamerImpl.java:75]: :thought_balloon: I think we might be using this in many places, should we put this in a util (e.g. BufferUtil) or as a default helper on `BufferWriter` itself, e.g. `BufferWriter#toByteArray`?

**Review by npepinpe — APPROVED** [review]
:+1: Great stuff, this is a big improvement on our previous iteration :slightly_smiling_face: 

Mostly suggestions and thoughts, didn't see any blockers.

One last thing, I dug this out of a branch I had where I tried to prototype some improvements for our futures. It might be useful in this case, but if you like it, I would propose we add it in a separate PR. I created this helper a while back but never merged it:

<details><summary>ActorFutureUtil.java</summary>

```java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Zeebe Community License 1.1. You may not use this file
 * except in compliance with the Zeebe Community License 1.1.
 */
package io.camunda.zeebe.scheduler.future;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.function.BiConsumer;

public final class ActorFutureUtil {

  /**
   * Awaits the completion of each future, one by one, in the order they are given. If all of them
   * are successful, the given callback is executed at the end with all null parameters. If one of
   * them fails, then the callback is immediately called with the error, and the next futures are
   * ignored.
   *
   * <p>NOTE: this helper is only usable from within an actor.
   *
   * @param callback the callback to call once on failure or once all futures are completed
   *     successfully
   * @param futures the queue of futures to await
   */
  public static void onComplete(
      final BiConsumer<Void, Throwable> callback, final ActorFuture<?>... futures) {
    final var queue = new ArrayDeque<>(Arrays.asList(futures));
    final var chain = new Chain(queue);
    chain.onComplete(callback);
  }

  private record Chain(Queue<ActorFuture<?>> futures) {

    private void onComplete(final BiConsumer<Void, Throwable> callback) {
      final ActorFuture<?> next = futures.poll();
      if (next == null) {
        callback.accept(null, null);
        return;
      }

      next.onComplete(
          (ok, error) -> {
            if (error != null) {
              callback.accept(null, error);
              futures.clear();
              return;
            }

            onComplete(callback);
          });
    }
  }
}
```

</details>

Then usage looks like this, say, for the `RemoteStreamServiceImpl#start`

```java
  @Override
  public ActorFuture<RemoteStreamer<M, P>> start() {
    // ... setup and so on

    ActorFutureUtil.onComplete(
        (ok, error) -> future.complete(streamer),
        actorSchedulingService.submit(streamer),
        actorSchedulingService.submit(apiServer),
        future);
    return future;
  }
```

**Pull request merged**